### ISSUE TITLE
Aaccuracy improvement

### ISSUE THREAD

**Issue by Rejea** [dbolya/yolact#204]
Is there any way to improve accuracy at the expense of speed？
And if i want to test the performance at different values of K, where should i make the change?

**Comment by Rejea** [comment]
What if i add one more FPN layer, P8, increasing num_downsample to 3,pred_aspect_ratios to '*6' and [12] in pred_scales.
Will it work?

**Comment by dbolya** [comment]
Yeah, that would work.

Other ways to tune the accuracy-speed curve would be to increase the image size (and to see how that works, look at the definition for `yolact_im700_config`).

Then to increase the number of prototypes, simply define a protonet with that many outputs, like this:
https://github.com/dbolya/yolact/blob/2d5bef153506c3a8313e2e3a043a64c939ca37ee/data/config.py#L660
The last layer in this definition is `(32, 1, {})` which means "32 output channels, kernel size of 1, and no extra arguments". To increase the number of prototypes, simply increase that 32 to something larger.

**Comment by Rejea** [comment]
WOW great thanks !!That was what i thought. But does it also mean i need to train it again?

And as a green hand in deep learning, i sincerely  want to ask for how to build P8.

**Comment by dbolya** [comment]
Yeah it woulds require retraining, since you're now adding new weights. You can build P8 exactly in the way that you described, but if your goal is to add a smaller scale, it might be more useful to generate P2 rather than P8 (since P2 will have features for smaller objects than P3 does). This will add a gazillion more anchors though, so it would slow it down a lot.

If you wanted to go the P2 route, you would do all the same steps you did for P8 but instead of changing `num_downsample`, you would change this selected layers:
https://github.com/dbolya/yolact/blob/2d5bef153506c3a8313e2e3a043a64c939ca37ee/data/config.py#L641 to include 0 (so make it `range(4)` instead of `range(1, 4)`). You can check the stats for each layer by running `python yolact.py <config_name>` to make sure you've done everything correctly.

**Comment by Rejea** [comment]
So it is to say, P8 will not help as much as P2 for small objects but slow down a little while P2 will slow down a lot.
When i try to retrain it ,i found this problem in my image. Do you have any ideas? Because i cannot fix it even after google.
![image](https://user-images.githubusercontent.com/57441019/68555954-2d862800-046b-11ea-8a97-8570a59dc5e9.png)

And i have a new confusion. When i used P3-P7, the original model, if the things in test image are small like six or seven oranges, each orange will be detected. And these six or seven oranges will also be detected as a whole. But this problem won't appear when i decrease num_downsample to 1 or 0 ( i think 1 means not using P7 and 0 means not using both P7and P6.). Does this mean less layers could have a better performance when objects are small and gathering in a cluster?

**Comment by Rejea** [comment]
I called it a  "detecting" problem because i wonder wether it is a localization failure. The objects are not covered but gathered.

e.g.
![806e76330cf0fabd1e4d71660ec657f5](https://user-images.githubusercontent.com/57441019/68556831-c66a7280-046e-11ea-9f94-8769b4943d19.jpg)

**Comment by Rejea** [comment]
Dear dbolya, sorry for having too many questions.

It is not difficult to understand each layer,Pi, gets three kinds of coefficients(class, box and mask) through prediction head. But how to get the final mask coefficients? Adding them together or doing something else.

To make me clear, i already have coefficients from Pi calld class_co_i, box_co_i and mask_co_i. What should i do next. And i remember, you use x to represent the output ,which is Pi in my example, in your code.

**Comment by abhigoku10** [comment]
@Rejea  by doing this your going to increase the weight of the model and improve the accuracy right ??  @dbolya  can we increase the layers and obtain heavier model which can detect smaller objects also at the expanse of speed

**Comment by dbolya** [comment]
Oh whoops I didn't see any of these replies so sorry for the delay.

From the top for @Rejea:
 - Yeah P2 would be better for small objects theoretically.
 - About your training issue: hmm, you might have configured something improperly. To sanity check, can you train with the base configs normally? Or do you also get the error there?
 - The big detection not appearing because you removed the later P layers is simply due to removing the layer the detection came from. An overlap detection like that could come from any layer, so removing the deepest ones wouldn't improve in general across all images.
 - The construct the masks from the masks coefficients, you simply take the prototypes from the output of protonet (which is independent of the prediction heads) and do `sigmoid(prototypes @ coefficients.t()` where @ is matrix multiplication. Check out the paper for explicit details and make sure your coefficients have run into a tanh and prototypes into a relu by this point.

Then @abhigoku10, you can and I think that's what @Rejea was trying to do. You should be able to do the same by following the steps above.

**Comment by Rejea** [comment]
@abhigoku10 yeah,that's what im trying to do. 

@dbolya GOT IT .
 And i met that training issue even if i use the original configs. that's what makes me confused.

**Comment by dbolya** [comment]
@Rejea Hmm what's your pytorch, torchvision, and python version?

Also did you do anything to disable grad computation manually like wrap something in torch.no_grad or disable it with an environment variable or something? Does training still not work even with a fresh clone of the current master branch?

**Comment by Rejea** [comment]
torch 1.2.0,torchvision0.4.0,python3.7.3
i think i could redownload the new version and test.

**Comment by dbolya** [comment]
YOLACT should work with those versions, so maybe yeah test with a clean version of the repo.

**Comment by moganesyan** [comment]
@Rejea Hey! Were you successful in increasing accuracy at the expense of speed? What worked best?

Thanks in advance!

**Comment by Rejea** [comment]
@moganesyan It depends actually. If you use it on small scales. P2 is better. But it slows down indeed.

Adding weights works too, but only improves a little.
